runtime/lua/lsp/client.lua

```diff
@@ -0,0 +1,391 @@
+local EmptyDictionary = require('nvim.meta').EmptyDictionary
+
+local message = require('lsp.message')
+local initialize_filetype_autocmds = require('lsp.autocmds').initialize_filetype_autocmds
+local lsp_doautocmd = require('lsp.autocmds').lsp_doautocmd
+local call_callbacks_for_method = require('lsp.callbacks').call_callbacks_for_method
+local should_send_message = require('lsp.checks').should_send
+
+local log = require('lsp.log')
+
+local read_state = Enum:new({
+  init = 0,
+  header = 1,
+  body = 2,
+})
+
+local error_level = Enum:new({
+  critical = 0,
+  reset_state = 1,
+  info = 2,
+})
+
+
+local ActiveJobs = {}
+
+ActiveJobs.add = function(id, obj)
+  ActiveJobs[id] = obj
+end
+
+ActiveJobs.remove = function(id)
+  ActiveJobs[id] = nil
+end
+
+local client = {}
+client.__index = client
+
+client.job_stdout = function(id, data)
+  if ActiveJobs[id] == nil then
+    return
+  end
+
+  ActiveJobs[id]:on_stdout(data)
+end
+
+client.job_exit = function(id, data)
+  if ActiveJobs[id] == nil then
+    return
+  end
+
+  ActiveJobs[id]:on_exit(data)
+end
+
+client.new = function(name, ft, cmd)
+  log.debug('Starting new client: ', name, cmd)
+
+  local job_id = vim.api.nvim_call_function('lsp#__jobstart', { cmd })
+
+  assert(job_id)
+  assert(job_id > 0)
+
+  log.trace('Client id: ', job_id)
+
+  local self = setmetatable({
+    job_id = job_id,
+
+    name = name,
+    ft = ft,
+    cmd = cmd,
+
+    -- State for handling messages
+    _read_state = read_state.init,
+    _read_data = '',
+    _current_header = {},
+
+    -- Capabilities sent by server
+    capabilities = EmptyDictionary:new(),
+
+    -- Results & Callback handling
+    --  Callbacks must take two arguments:
+    --      1 - success: true if successful, false if error
+    --      2 - data: corresponding data for the message
+    _callbacks = {},
+    _results = {},
+
+    -- Data fields, to be used internally
+    __data__ = {},
+  }, client)
+
+  ActiveJobs.add(job_id, self)
+
+  return self
+end
+
+client.initialize = function(self)
+  -- Initialize autocmd for filetypes == self.ft
+  --    These will create buffer local autocmds in each buffer
+  initialize_filetype_autocmds(self.ft)
+
+  local result = self:request_async('initialize', nil, function(_, data)
+    self:notify('initialized')
+    self:notify('textDocument/didOpen')
+    self.capabilities =  EmptyDictionary:new(data.capabilities)
+    return data.capabilities
+  end)
+
+
+  return result
+end
+
+client.close = function(self)
+  if self._closed then
+    return
+  end
+
+  self._closed = true
+  vim.api.nvim_call_function('jobclose', {self.job_id})
+end
+
+--- Make a request to the server
+-- @param method: Name of the LSP method
+-- @param params: the parameters to send
+-- @param cb (optional): If sent, will call this when it's done
+--                          otherwise it'll wait til the client is done
+client.request = function(self, method, params, cb)
+  if not method then
+    return nil
```

Comment:
How about using

```lua
error("No request method supplied", 2)
```
here? Not supplying a method name surely is an error, and this would provide a way to distinguish an empty reply from it.

Reply:
You could also remove that `return nil` now, won't ever be executed.